File: src/main/scala/cz/cvut/fit/cervamar/gremlin/core/ExecutorQuery.scala
package cz.cvut.fit.cervamar.gremlin.core

import javax.script.ScriptContext

import io.gatling.core.session.Session
import org.apache.tinkerpop.gremlin.driver.Client
import org.apache.tinkerpop.gremlin.groovy.jsr223.GremlinGroovyScriptEngine
import org.apache.tinkerpop.gremlin.structure.Graph

import scala.collection.JavaConverters._

/**
  * Created on 5/3/2017.
  *
  * @author Marek.Cervak
  */

trait QueryExecutor {
  def eval(gremlinQuery: GremlinQuery, session: Session) : Object
}
class ExecutorQuery(graph: Graph) extends QueryExecutor {

  import ExecutorQuery._
  val engine = new GremlinGroovyScriptEngine()
  engine.put(GRAPH, graph)
  engine.put(TRAVERSAL, graph.traversal())

  override def eval(gremlinQuery: GremlinQuery, session: Session): AnyRef = {
      val bindings = engine.getBindings(ScriptContext.ENGINE_SCOPE)
      //gremlinQuery.getAllVariables(session).foreach(x => bindings.put(x._1,x._2))
      gremlinQuery.getVariables.foreach(x => {
      bindings.put(x._1,x._2)
    })
      val plainQuery = gremlinQuery.getPlainQuery(session, graph.features.vertex.supportsNumericIds)
      println(plainQuery)
      engine.eval(plainQuery, bindings)
  }
}

object ExecutorQuery {
  val GRAPH = "graph"
  val TRAVERSAL = "g"
}

class ServerExecutor(client: Client) extends QueryExecutor {

  override def eval(gremlinQuery: GremlinQuery, session: Session): AnyRef = {
    client.submit(gremlinQuery.getPlainQuery(session, false), gremlinQuery.getAllVariables(session).asJava)
  }
}

File: src/test/scala/cz/cvut/fit/cervamar/gremlin/core/ExecutorQuery.java
package cz.cvut.fit.cervamar.gremlin.core;

import javax.script.Bindings;
import javax.script.ScriptContext;
import javax.script.ScriptEngine;
import javax.script.ScriptException;

import io.gatling.core.session.Session;
import lombok.extern.slf4j.Slf4j;
import org.apache.tinkerpop.gremlin.groovy.jsr223.GremlinGroovyScriptEngine;
import org.apache.tinkerpop.gremlin.structure.Graph;

/**
 * Created on 1/6/2018.
 *
 * @author Marek.Cervak
 */
@Slf4j
public class ExecutorQuery {

    private static final String GRAPH = "graph";
    private static final String TRAVERSAL = "g";
    private final boolean supportNumericIds;
    private ScriptEngine engine = new GremlinGroovyScriptEngine();


    public ExecutorQuery(Graph graph, boolean supportNumericIds) {
        this.supportNumericIds = supportNumericIds;
        engine.put(GRAPH, graph);
        engine.put(TRAVERSAL, graph.traversal());
    }

    Object eval(GremlinQuery gremlinQuery, Session session) throws ScriptException {
        ResolvedQuery resolvedQuery = gremlinQuery.getResolvedQuery(session, supportNumericIds);
        Bindings bindings = engine.getBindings(ScriptContext.ENGINE_SCOPE);
        resolvedQuery.bindings().forEach(bindings::put);
        log.debug("Query {} with bindings {}", resolvedQuery.query(), resolvedQuery.bindings());
        return engine.eval(resolvedQuery.query(), bindings);
    }
}
